trinity/utils/datastructures.py

```diff
@@ -286,3 +297,315 @@ def complete(self, batch_id: int, completed: Tuple[TTask, ...]) -> None:
+        """
+        if len(self._tasks) > 0 or self._oldest_depth != 0:
+            raise ValidationError(f"Tasks already added, cannot set {finished_task!r} as finished")
+
+        completed = self._prereq_tracker(finished_task)
+        completed.set_complete()
+        task_id = self._id_of(finished_task)
+        self._tasks[task_id] = completed
+        self._depths[task_id] = 0
+        # note that this task is intentionally *not* added to self._unready
+
+    def register_tasks(self, tasks: Tuple[TTask]) -> None:
+        """
+        Initiate a task into tracking. Each task must be registered *after* its dependency has
+        been registered.
+
+        :param tasks: the tasks to register, in iteration order
+        """
+        task_meta_info = (
+            (self._prereq_tracker(task), self._id_of(task), self._dependency_of(task))
+            for task in tasks
+        )
+
+        for prereq_tracker, task_id, dependency_id in task_meta_info:
+            if dependency_id not in self._tasks:
+                raise ValidationError(
+                    f"Cannot prepare task {prereq_tracker!r} with id {task_id} and "
+                    f"dependency {dependency_id} before preparing its dependency"
+                )
+            else:
+                self._tasks[task_id] = prereq_tracker
+                self._unready.add(task_id)
+                self._dependencies[dependency_id].add(task_id)
+                depth = self._depths[dependency_id] + 1
+                self._depths[task_id] = depth
+
+    def finish_prereq(self, subtask: TSubtask, tasks: Tuple[TTask]) -> None:
+        """For every task in tasks, mark the given prerequisite as completed"""
+        if len(self._tasks) == 0:
+            raise ValidationError("Cannot finish a task until set_last_completion() is called")
+
+        for task in tasks:
+            task_id = self._id_of(task)
+            if task_id not in self._tasks:
+                raise ValidationError(f"Cannot finish task {task_id!r} before preparing it")
+
+            task_completion = self._tasks[task_id]
+            task_completion.finish(subtask)
+            if task_completion.is_complete and self._dependency_of(task) not in self._unready:
+                self._mark_complete(task_id)
+
+    async def ready_tasks(self) -> Tuple[TTask, ...]:
+        """
+        Return the next batch of tasks that are ready to process. If none are ready,
+        hang until at least one task becomes ready.
+        """
+        queue = self._ready_tasks
+        if queue.empty():
+            first_task = await queue.get()
+        else:
+            first_task = queue.get_nowait()
+
+        # In order to return from get() as soon as possible, never await again.
+        # Instead, take only the tasks that are already available.
+        available = queue.qsize()
+
+        available_tasks = tuple(queue.get_nowait() for _ in range(available))
+
+        completed = (first_task, ) + available_tasks
```

Comment:
I feel like this pattern is present in multiple places,  seems like a good candidate for a utility function.